src/core/lib/surface/completion_queue.c

```diff
@@ -643,8 +643,7 @@ static void cq_end_op_for_next(grpc_exec_ctx *exec_ctx,
   /* Add the completion to the queue */
   bool is_first = cq_event_queue_push(&cqd->queue, storage);
   gpr_atm_no_barrier_fetch_add(&cqd->things_queued_ever, 1);
-  bool will_definitely_shutdown =
-      gpr_atm_no_barrier_load(&cqd->pending_events) == 1;
+  bool will_definitely_shutdown = gpr_atm_acq_load(&cqd->pending_events) == 1;
```

Comment:
Again, relating this issue to the bug report: which operation below is required to have a happens before ordering here? The later update of pending_events seems to have a full memory fence already, so I'm not sure that I'm seeing it here.

Reply:
T1 wrote to the `cq` via  `cq_event_queue_push` in line 644 which is a lock free operation (and it did without having completion queue lock or without any other memory fence ) .. 

T2 called `completion_queue_destroy` which called cq-shutdown first which decremented `pending_events` to zero (under cq-lock AND also under a memory fence by doing `full_fetch_add` i.e but that doesn't matter) and this ended up destroying the `cq`

My claim was that since there was no happens-before between T1 and T2,  tsan is complaining that there is a data race (It says T2 did a write and the previous write is done by T1 )

This line is really the only fix to this bug.  However, I felt it is a good practice to just use `acq` for load and `release` for store as a good practice in general (especially considering the fact that using `no_barrier` operations was not a deliberate choice in the first place :) )..

Reply:
But how is this introducing a happens-before between T1 and T2 as you described? The function cq_event_push works on various variables but not the variables that you're touching here, so I'm not seeing the needed conflict order edge. The F&A involved in the TSAN race is on `cqd->queue.num_queue_items` but the freeing is happening because of the result of a F&A on `cqd->pending_events`.

Reply:
Put another way: is the real signal that it's safe to free that both `pending_events` and `num_queue_items` are at 0?

Reply:
When you do an "acquire load" on `pending_events` in T1 on line 646 , that will "match" will the release store (rather the full cas) on `pending_events` done by T2 - thereby establishing T2 happened before T1 .